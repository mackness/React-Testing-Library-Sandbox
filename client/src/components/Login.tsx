import React, { SyntheticEvent } from "react";
import { Mutation, ApolloConsumer } from "react-apollo";
import gql from "graphql-tag";

import styled from "@emotion/styled";

import LoginForm from "./LoginForm";

export const LOGIN_USER = gql`
  mutation login($email: String!) {
    login(email: $email)
  }
`;

interface LoginProps {
  submitHandler: (event: any, login: Function) => void;
}

function submitHandler(event: any, login: Function): void {
  login({
    variables: { email: event.target.querySelector("input[type=text]").value }
  });
}

export default function Login({ submitHandler }: LoginProps): JSX.Element {
  return (
    <ApolloConsumer>
      {client => {
        return (
          <Mutation
            mutation={LOGIN_USER}
            onCompleted={({ login }: any) => {
              localStorage.setItem("token", login);
              client.writeData({ data: { isLoggedIn: true } });
            }}
          >
            {(login: any, { loading, error }: any) => {
              if (loading) return <p>loading...</p>;
              if (error) return <p>an error occurred...</p>;
              return (
                <>
                  <Title>Pok√®dex</Title>
                  <LoginForm
                    submitHandler={(event: SyntheticEvent) =>
                      submitHandler(event, login)
                    }
                  />
                </>
              );
            }}
          </Mutation>
        );
      }}
    </ApolloConsumer>
  );
}

Login.defaultProps = {
  submitHandler
};

const Title = styled("h3")({
  marginBottom: 8
});
