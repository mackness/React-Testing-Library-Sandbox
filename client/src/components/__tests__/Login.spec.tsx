import "react-testing-library/cleanup-after-each";
import * as React from "react";
import { createSerializer } from "jest-emotion";
import * as emotion from "emotion";
import { MockedProvider } from "react-apollo/test-utils";
import { render } from "react-testing-library";
import Login, { LOGIN_USER } from "../Login";

expect.addSnapshotSerializer(createSerializer(emotion));

let mocks: any;
beforeEach(() => {
  mocks = [
    {
      request: {
        query: LOGIN_USER,
        variables: {
          email: "m@te.co"
        }
      },
      error: null,
      result: {
        data: {
          login: (k: any) => k
        }
      }
    }
  ];
});

jest.mock("../LoginForm", () => () => <span>login form</span>);

describe("components", () => {
  describe("Login", () => {
    it("should mount without crashing", () => {
      const { container } = render(
        <MockedProvider mocks={mocks}>
          <Login />
        </MockedProvider>
      );
      expect(container).toBeDefined();
    });

    it("should render the title with the correct text content", () => {
      const { getByText } = render(
        <MockedProvider mocks={mocks}>
          <Login />
        </MockedProvider>
      );
      expect(getByText(/PokÃ¨dex/i)).toBeDefined();
    });

    // as a note this snapshot test should contain human readable css properties
    // thanks to the emotion snapshot serializer
    it("should pass snapshot test", () => {
      const { container } = render(
        <MockedProvider mocks={mocks}>
          <Login />
        </MockedProvider>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
