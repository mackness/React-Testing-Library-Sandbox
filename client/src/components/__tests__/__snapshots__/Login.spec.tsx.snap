// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Login should pass snapshot test 1`] = `
.emotion-0 {
  margin-bottom: 8px;
}

<div>
  <h3
    class="emotion-0"
  >
    Pok√®dex
  </h3>
  <span>
    login form
  </span>
</div>
`;
